### PR TITLE
[utils] Add a preliminary expression parser in eos and its test.

### PR DESCRIPTION
**Pull request by mreboud**
This is still preliminary work since Kinematics are not yet fully implemented.

### CODE REVIEW

File: eos/utils/expression-fwd.hh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+#include <eos/utils/one-of.hh>
+
+namespace eos::exp
+{
+    class BinaryExpression;
+    class ConstantExpression;
+    class ObservableNameExpression;
+    class ObservableExpression;
+
+    typedef OneOf<BinaryExpression, ConstantExpression, ObservableNameExpression, ObservableExpression> Expression;
```
Comment (dvandyk) [inline eos/utils/expression-fwd.hh:30]: Please use ``using Expression = OneOf<...>;``.

File: eos/utils/expression-maker.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+#include <iostream>
+
+namespace eos::exp
+{
+
+    // copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
+    class ExpressionMaker
+    {
+        Parameters _parameters;
```
Comment (dvandyk) [inline eos/utils/expression-maker.cc:34]: Please add an explicit ``private:`` and indent the members one level.

File: eos/utils/expression-maker.cc [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    )
+                );
+            }
+
+            Expression visit(ObservableExpression &)
+            {
+                throw InternalError("Encountered ObserableExpression in ExpressionMaker::visit");
+            }
+    };
+}
```
Comment (dvandyk) [inline eos/utils/expression-maker.cc:72]: Missing newline at the end of the file.

File: eos/utils/expression-parser-impl.hh [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                )
+            ;
+
+        primary_expr =
+              ( '(' >> expression >> ')' )               [ _val = _1 ]
+            | constant                                   [ _val = _1 ]
+            | observable_name                            [ _val = _1 ]
+            ;
+
+        constant = double_ | int_;
```
Comment (dvandyk) [inline eos/utils/expression-parser-impl.hh:61]: I guess we will not need complex-valued numbers, since this works at the level of observables. Or am I wrong?

File: eos/utils/expression-parser.hh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+        // Constructor
+        ExpressionParser();
+
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> expression;
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> additive_expr;
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> non_additive_expr;
+
+        qi::rule<Iterator, eos::exp::Expression()               , ascii::space_type> primary_expr;
+        qi::rule<Iterator, eos::exp::ConstantExpression()       , ascii::space_type> constant;
```
Comment (dvandyk) [inline eos/utils/expression-parser.hh:62]: Why do you specialize here to ``eos::exp::ConstantExpression``? I would leave the more general ``eos::exp::Expression``.

File: eos/utils/expression-parser.hh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        qi::rule<Iterator, eos::exp::ObservableNameExpression() , ascii::space_type> observable_name;
```
Comment (dvandyk) [inline eos/utils/expression-parser.hh:63]: Ditto.

File: eos/utils/expression-parser_TEST.cc [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+#include <eos/utils/parameters.hh>
+
+#include <iostream>
+
+using namespace test;
+using namespace eos::exp;
+using namespace eos;
+
+
+class ExpressionEvaluator
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:36]: This should be moved somewhere else.

File: eos/utils/expression-parser_TEST.cc [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            throw InternalError("Encountered ObserableNameExpression in ExpressionEvaluator::visit");
+
+            return 0.0;
+        }
+};
+
+
+struct ExpressionTest
+{
+    const std::string _input;
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:67]: The leading underscores indicate that this (and all further members) are private. But they are not. Please remove the leading underscores or make them private.

File: eos/utils/expression-parser_TEST.cc [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+{
+public:
+    ExpressionParserTest() :
+    TestCase("expression_parser_test")
+    {
+    }
+
+    virtual void run() const
+    {
+        {
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:96]: Please a a comment before this scope, explaining briefly what's being tested. E.g.,
```C++
// testing basic parsing of constants and binary expressions
```
or some such.

File: eos/utils/expression-parser_TEST.cc [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+            ExpressionTest test("1+2*3");
+
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test._e.accept(printer);
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:101]: Here the naming rang my alarm bells, since this looks like you're using a private member variable.

File: eos/utils/expression-parser_TEST.cc [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test._completed);
+            TEST_CHECK_EQUAL(test._e.accept_returning<double>(evaluator), 7);
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:106]: ``7.0``

File: eos/utils/expression-parser_TEST.cc [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+            TEST_CHECK_THROWS(InternalError, test2._e.accept_returning<double>(evaluator));
+
+            Parameters p = Parameters::Defaults();
+            Kinematics k = Kinematics({{"q2_min", 0.011164}, {"q2_max", 11.62}});
+            Options o;
+            ExpressionMaker maker(p, k, o);
+            Expression evaluable_test2 = test2._e.accept_returning<Expression>(maker);
+
+            // This test is to be improved with correct kinematics and default parameters
+            TEST_CHECK_RELATIVE_ERROR(evaluable_test2.accept_returning<double>(evaluator), 0.295627626, 1e-5);
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:134]: Hm, this provide a circular dependence on the EOS libraries, but only when testing: ``libeosutils`` must be built before ``libeosbdecays.so``, but the tests of ``libeosutils`` depend on ``libeosbdecays``. Since it works, let's be happy and leave it as is.

File: eos/utils/expression.cc [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+ * Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <eos/utils/expression.hh>
+
+//#include <iostream>
+
+namespace eos::exp
+{
+    double BinaryExpression::Add(double a, double b)      { return a+b; }
```
Comment (dvandyk) [inline eos/utils/expression.cc:24]: Please use lower case for all members functions (even static), except for one very particular case: named constructors.
This is not one, so please convert to lower case.

File: eos/utils/expression.cc [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    double BinaryExpression::Subtract(double a, double b) { return a-b; }
+    double BinaryExpression::Multiply(double a, double b) { return a*b; }
+    double BinaryExpression::Divide(double a, double b)   { return a/b; }
+
+    BinaryExpression::func
+    BinaryExpression::Method(char op)
```
Comment (dvandyk) [inline eos/utils/expression.cc:30]: Ditto.

File: eos/utils/expression.cc [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+#include <eos/utils/expression.hh>
+
+//#include <iostream>
+
+namespace eos::exp
+{
+    double BinaryExpression::Add(double a, double b)      { return a+b; }
+    double BinaryExpression::Subtract(double a, double b) { return a-b; }
+    double BinaryExpression::Multiply(double a, double b) { return a*b; }
+    double BinaryExpression::Divide(double a, double b)   { return a/b; }
```
Comment (dvandyk) [inline eos/utils/expression.cc:27]: Come to think of it: ``add``, ``subtract``, ``multiply`` and ``divide`` are not expressions, they are actions. The expressions are ``sum``, ``difference``, ``product`` and ``ratio``. Care to change that?

File: eos/utils/expression.cc [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+    BinaryExpression::func
+    BinaryExpression::Method(char op)
+    {
+        switch(op) {
+            case '+': return BinaryExpression::Add;
+            case '-': return BinaryExpression::Subtract;
+            case '*': return BinaryExpression::Multiply;
+            case '/': return BinaryExpression::Divide;
+            default:  return nullptr;
```
Comment (dvandyk) [inline eos/utils/expression.cc:37]: This should not happen, should it? I would throw an
```C++
eos::InternalError("Unknown binary operator '" + stringify(op) + "' encountered");
```

File: eos/utils/expression.hh [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace eos::exp
+{
+    class BinaryExpression
+    {
+        public:
+            char _op;
+            Expression _lhs, _rhs;
+
+            typedef double(*func)(double, double);
```
Comment (dvandyk) [inline eos/utils/expression.hh:37]: Please use ``using``:
```C++
using func = (double)(*)(double, double);
```

Why are you not using ``const double &`` for the two arguments?

File: eos/utils/expression.hh [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+            static double Add(double a, double b);
```
Comment (dvandyk) [inline eos/utils/expression.hh:39]: See comment above: ``sum`` vs ``add``, etc.

File: eos/utils/expression.hh [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+            static double Subtract(double a, double b);
+            static double Multiply(double a, double b);
+            static double Divide(double a, double b);
+
+            static BinaryExpression::func Method(char op);
+
+            BinaryExpression() {}
+
+            BinaryExpression(char op, Expression const& l, Expression const& r) :
```
Comment (dvandyk) [inline eos/utils/expression.hh:48]: EOS coding style puts ``const`` before the reference's type, i.e.: ``const Expression & l``, with all the spaces.

File: eos/utils/expression.hh [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+                _op(op), _lhs(l), _rhs(r)
+            {
+            }
+    };
+
+    class ConstantExpression
+    {
+        public:
+            double _value;
```
Comment (dvandyk) [inline eos/utils/expression.hh:57]: Pleasr remove the leading underscore.

File: eos/utils/expression.hh [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+#include <memory>
+#include <iostream>
+#include <map>
+
+namespace eos::exp
+{
+    class BinaryExpression
+    {
+        public:
+            char _op;
```
Comment (dvandyk) [inline eos/utils/expression.hh:34]: Please remove the leading underscores.

File: eos/utils/expression.hh [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+            {
+            }
+    };
+
+    class ObservableNameExpression
+    {
+        public:
+            std::string _observable_name;
+
+            ObservableNameExpression(const std::string& input = "") :
```
Comment (dvandyk) [inline eos/utils/expression.hh:70]: space between ``string`` and ``&``

File: eos/utils/expression.hh [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+                 _observable_name(input)
+            {
+            }
+
+    };
+
+    class ObservableExpression
+    {
+        public:
+            ObservablePtr _observable;
```
Comment (dvandyk) [inline eos/utils/expression.hh:80]: Underscore.

File: eos/utils/expression.hh [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+            ConstantExpression(const double& v = 0) :
+                _value(v)
+            {
+            }
+    };
+
+    class ObservableNameExpression
+    {
+        public:
+            std::string _observable_name;
```
Comment (dvandyk) [inline eos/utils/expression.hh:68]: Underscore.

**Review by dvandyk — CHANGES_REQUESTED** [review]
Thank you Meril, well done! It looks very good! I have a few comments, nothing serious.

File: eos/utils/expression-parser-impl.hh (hunk 4 above)
Comment (mreboud) [reply eos/utils/expression-parser-impl.hh:61]: This is something we can easily add if needed. I would keep the parser as simple as possible until we have a complete working basis if you agree.

File: eos/utils/expression-parser_TEST.cc (hunk 11 above)
Comment (mreboud) [reply eos/utils/expression-parser_TEST.cc:106]: In principle, this is parsed as integers, so the result is `(int) 7`. In practice, I am happy with `7.0` because `TEST_CHECK_EQUAL` doesn't test the type :)

Comment (mreboud) [reply eos/utils/expression-parser_TEST.cc:106]: Ah no sorry, my mistake, the evaluator evaluates to `double`.

Comment (dvandyk) [reply eos/utils/expression-parser_TEST.cc:106]: Hm, but your visitor can only return a double!

File: eos/utils/expression-parser-impl.hh (hunk 4 above)
Comment (dvandyk) [reply eos/utils/expression-parser-impl.hh:61]: I do! No complex values for now.

File: eos/utils/expression-parser.hh (hunk 5 above)
Comment (mreboud) [reply eos/utils/expression-parser.hh:62]: It simply doesn't work with `eos::exp::Expression`. I think this is needed to construct the correct class directly (with `_val=_1`).

**Comment by mreboud** [comment]
> Thank you Meril, well done! It looks very good! I have a few comments, nothing serious.

Thanks Danny for the review; I accounted for all of your comments but those concerning class specialization in the parser.

File: eos/utils/expression-parser.hh (hunk 5 above)
Comment (dvandyk) [reply eos/utils/expression-parser.hh:62]: Understood. Keep it as is then.

**Comment by dvandyk** [comment]
@mreboud Can you please include my two commits for aliasing kinematic variables (``gh-dvandyk/for-meril``) in this PR?

**Review by dvandyk — APPROVED** [review]

File: eos/utils/kinematic.cc [hunk 25]
```diff
@@ -153,39 +174,86 @@ namespace eos
             result += ", " + i->first + '=' + stringify(_imp->variables_data[i->second]);
         }
 
+        // TODO: show aliases
```
Comment (dvandyk) [inline eos/utils/kinematic.cc:177]: Please remove this. We should not show the aliases at all.

File: eos/utils/kinematic.cc [hunk 26]
```diff
@@ -153,39 +174,86 @@ namespace eos
...
+        const auto j(_imp->alias_map.find(name));
+        bool name_exists  = (_imp->variables_map.end() != i);
+        bool alias_exists = (_imp->alias_map.end() != j);
+
+        if (! name_exists)
+            throw UnknownKinematicVariableError(name);
+
+        if (alias_exists)
+        {
+            // TODO: proper error
```
Comment (dvandyk) [inline eos/utils/kinematic.cc:195]: Can you please create a ``DuplicateKinematicAliasError`` in the same place where ``UnknownKinematicVariableError`` lives, and use it here? I was lazy.

File: eos/utils/kinematic.cc [hunk 27]
```diff
@@ -200,9 +268,10 @@ namespace eos
         return _imp->variables.end();
     }
 
-    KinematicVariable::KinematicVariable(const std::shared_ptr<Implementation<Kinematics>> & imp, unsigned index) :
+    KinematicVariable::KinematicVariable(const std::shared_ptr<Implementation<Kinematics>> & imp, unsigned index, bool is_alias) :
         _imp(imp),
-        _index(index)
+        _index(index),
+        _is_alias(is_alias)
```
Comment (dvandyk) [inline eos/utils/kinematic.cc:274]: I don't think ``_is_alias`` is required anymore now. However, I'm happy to leave it in, it might be useful later and does it does not put pressure on memory consumption.

File: eos/utils/kinematic_TEST.cc [hunk 28]
```diff
@@ -48,6 +48,30 @@ class KinematicsTest :
...
+                TEST_CHECK_EQUAL(1.0, kinematics["s_min"]);
+                TEST_CHECK_EQUAL(6.0, kinematics["s_max"]);
+                TEST_CHECK_EQUAL(1.0, kinematics["q2_min"]);
+                TEST_CHECK_EQUAL(6.0, kinematics["q2_max"]);
+
+                TEST_CHECK_EQUAL("s_min", kinematics["s_min"].name());
+                TEST_CHECK_EQUAL("s_min", kinematics["s_min"].name());
+
+                // Test clearing the alias map and access values
+                auto saved_alias = kinematics["s_min"];
```
Comment (dvandyk) [inline eos/utils/kinematic_TEST.cc:70]: ``s_min`` is not an alias, it is **aliased*. Please add a test case that saves, in addition, ``q2_min`` and ``q2_max`` and check that they surive clearing the alias map.

File: eos/utils/expression-maker.cc [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+                    );
+                // Clear alias map
+                _kinematics.clear_aliases();
+                // Return observable
+                return ObservableExpression(observable);
+            }
+
+            Expression visit(ObservableExpression &)
+            {
+                throw InternalError("Encountered ObserableExpression in ExpressionMaker::visit");
```
Comment (dvandyk) [inline eos/utils/expression-maker.cc:87]: Hm, is this correct? Shouldn't this simply pass through? I don't think we have any need to apply this visitor twice in a row, but it would be perfectly valid, wouldn't it?

File: eos/utils/expression-parser-impl.hh [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+              '{'
+            >> as_string [ lexeme [ *~char_("}") ] ]     [ _val = _1 ]
+            >> '}'
+            ;
+
+        kinematics =
+            '['
+            >> (
+                kinematics_alias    [ phx::bind(_val, _1) ]
+              | kinematics_value    [ phx::bind(_val, _1) ]
```
Comment (dvandyk) [inline eos/utils/expression-parser-impl.hh:83]: I failed to indent this properly w.r.t. to the other attributes in e.g. line 75.

File: eos/utils/expression-parser_TEST.cc [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+            std::stringstream out;
+            ExpressionPrinter printer(out);
+            test2.e.accept(printer);
+
+            ExpressionEvaluator evaluator;
+
+            TEST_CHECK(test2.completed);
+            TEST_CHECK_EQUAL_STR(
+                "BinaryExpression(ObservableNameExpression(B->Dlnu::BR;l=tau, aliases=[q2_min=>q2_min_tau])\
+ / ObservableNameExpression(B->Dlnu::BR;l=mu, values=[q2_min=0]))",
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:96]: I prefer a different multiline C string:
```
TEST_CHECK_EQUAL_STR(
    "BinaryExpression(ObservableNameExpression(B->Dlnu::BR;l=tau, aliases=[q2_min=>q2_min_tau])"
    " / "
    "ObservableNameExpression(B->Dlnu::BR;l=mu, values=[q2_min=0]))",
    out.str()
);
```

File: eos/utils/expression-printer.cc [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+                        _os << "," << v->first << "=" << v->second;
+                    }
+                    _os << "]";
+                }
+                _os << ")";
+            }
+
+            void visit(ObservableExpression &)
+            {
+                throw InternalError("Encountered ObserableExpression in ExpressionPrinter::visit");
```
Comment (dvandyk) [inline eos/utils/expression-printer.cc:81]: Is this correct? We might want to print an expression that includes observables!

**Review by dvandyk — CHANGES_REQUESTED** [review]
Some tiny changes, but some more important are:
  - questions regarding the treatment of ``ObservableExpression`` in ``ExpressionEvaluator`` and ``ExpressionPrinter``;
  - incorrect test case when clearing a ``Kinematics``'s aliases.

**Comment by mreboud** [comment]
There are still a couple of steps before we can merge:
- [ ]  Fill the `_kinematics_names_array` of the class `ExpressionObservableEntry`: can be done via a visitor method.
- [ ] Implement the `uses(...)` in the class `ExpressionObservable`
- [ ] Check that object, esp. of the type `Expression`, are cloned when needed and shared when possible

File: eos/utils/expression-cloner.cc [hunk 33]
```diff
@@ -0,0 +1,90 @@
...
+            }
+
+            Expression visit(const ConstantExpression & e)
+            {
+                return ConstantExpression(e);
+            }
+
+            Expression visit(const ObservableExpression & e)
+            {
+                auto kinematics_values = e.kinematics_specification.values;
```
Comment (dvandyk) [inline eos/utils/expression-cloner.cc:59]: I would use ``const auto &`` here.

File: eos/utils/expression-cloner.cc [hunk 34]
```diff
@@ -0,0 +1,90 @@
...
+#include <eos/utils/options.hh>
+#include <eos/utils/parameters.hh>
+#include <eos/utils/qualified-name.hh>
+
+#include <iostream>
+
+namespace eos::exp
+{
+
+    // copies one expression tree to a new tree, replacing ObservableNameExpression objects with ObservableExpression objects
```
Comment (dvandyk) [inline eos/utils/expression-cloner.cc:31]: This comment does not reflect what the visitor does.

File: eos/utils/expression-maker.cc [hunk 35]
```diff
@@ -0,0 +1,90 @@
...
+
+                return ObservableExpression(observable, e.kinematics_specification);
+            }
+
+            Expression visit(const ObservableExpression & e)
+            {
+                return e;
+            }
+    };
+}
```
Comment (dvandyk) [inline eos/utils/expression-maker.cc:90]: ``ExpressionCloner`` and ``ExpressionMaker`` seem to be very similar. Can we combine them into one class?

File: eos/utils/expression-observable.hh [hunk 36]
```diff
@@ -0,0 +1,193 @@
...
+                    const Kinematics & kinematics,
+                    const Options & options,
+                    const Expression & expression) :
+                _name(name),
+                _parameters(parameters),
+                _kinematics(kinematics),
+                _options(options),
+                _expression(expression)
+            {
+                // Uses to be implemented
```
Comment (dvandyk) [inline eos/utils/expression-observable.hh:65]: Should we make a ``ParameterUser`` a ``ExpressionVisitor``? That would be an elegant solution to implementing parameter usage.

**Review by dvandyk — COMMENTED** [review]
A few comments...

File: eos/utils/expression-maker.cc (hunk 35 above)
Comment (mreboud) [reply eos/utils/expression-maker.cc:90]: Good point

Comment (mreboud) [reply eos/utils/expression-maker.cc:97]: All well thought, this is maybe not a good idea. The `maker` works on `ObservableNameExpression` while the `cloner` focuses on already made `ObservableExpression`. What I can do is gather the visitors into one `expression-visitors.cc` to avoid having so many files.

File: eos/utils/expression-parser-impl.hh [hunk 37]
```diff
@@ -0,0 +1,110 @@
...
+            | constant                                   [ _val = _1 ]
+            | (observable_name >> kinematics)            [ _val = phx::bind(make_observable, _1, _2) ]
+            | observable_name                            [ _val = phx::bind(make_observable, _1, KinematicsSpecification()) ]
+            ;
+
+        constant = double_ | int_;
+        observable_name =
+              '{'
+            >> as_string [ lexeme [ *~char_("}") ] ]     [ _val = _1 ]
+            >> '}'
```
Comment (dvandyk) [inline eos/utils/expression-parser-impl.hh:76]: Hm, we have a (small) problem here. What about
``b->smumu::Re{c9}``? We should probably:
 - use ``{{`` and ``}}`` as lead in and out for an observable name,
 - prohibit ``{{`` and ``}}`` anywhere inside a ``QualifiedName``.

**Review by dvandyk — CHANGES_REQUESTED** [review]

File: eos/utils/expression-cloner.cc [hunk 38]
```diff
@@ -0,0 +1,88 @@
...
+                    _kinematics.declare(value.first, value.second);
+                }
+                for (const auto & alias : kinematics_aliases)
+                {
+                    _kinematics.alias(alias.first, alias.second);
+                }
+
+                // Make observable
+                auto observable = Observable::make(
+                    e.observable->name().full(),
```
[the remark this answers is not in the document]
Comment (mreboud) [reply eos/utils/expression-cloner.cc:77]: Maybe I misunderstand `full` vs `str`, but I want to be sure that all options specified in the `QualifiedName` are kept in the process.

File: eos/utils/expression-parser-impl.hh (hunk 37 above)
Comment (dvandyk) [reply eos/utils/expression-parser-impl.hh:76]: As pointed out by Meril on Discord: there are Observables that use ``}}`, hence we have to adjust. Using ``<<`` and ``>>`` now as delimiters.

Comment (mreboud) [reply eos/utils/expression-parser-impl.hh:76]: I will add a commit to prohibit `<<` and `>>` in the prefix, the rest is already forbidden.

File: eos/utils/expression-maker.cc (hunk 35 above)
Comment (dvandyk) [reply eos/utils/expression-maker.cc:98]: Understood!

File: python/_eos.cc [hunk 39]
```diff
@@ -298,6 +299,30 @@ BOOST_PYTHON_MODULE(_eos)
         .def("__str__", &Options::as_string)
         ;
 
+    // Units
+    class_<Unit, boost::noncopyable>("Unit", R"(
+            Represents the unit of the observables.
+
+            Seven possible entries are currently implemented in EOS:
+            Undefined, None, GeV, GeV2, InverseGeV2, InverseGeV4, InversePicoSecond.
```
Comment (dvandyk) [inline python/_eos.cc:307]: Can you make this comment a list? It will be rendered as such in the documentation. You can check the documentation locally with
```
make install
make -C doc html
```
The doc will be available as ``doc/_build/html/index.html``.

File: python/_eos.cc [hunk 40]
```diff
@@ -589,6 +614,18 @@ BOOST_PYTHON_MODULE(_eos)
...
         .def("__getitem__", &Observables::operator[])
         .def("__iter__", range(&Observables::begin, &Observables::end))
+        .def("insert", &Observables::insert, R"(
+            Insert a new observable to EOS by parsing the input string.
+
+            :param name: The name of the new observable.
+            :type name: eos.QualifiedName
+            :param latex: The latex representation of the new observable.
+            :type latex: std::string
+            :param options: The set of options relevant to this new observable. These options apply to all the observables in the expression.
```
Comment (dvandyk) [inline python/_eos.cc:624]: Can you emphasize the 'all' here for the documentation, i.e., ``**all**``? Generally, we can use all the markup available in reStructuredText, which is also what we're using for the parameter descriptions.

File: eos/observable.cc [hunk 41]
```diff
@@ -86,7 +88,11 @@ namespace eos
 
             void insert(const QualifiedName & key, const std::shared_ptr<const ObservableEntry> & value)
             {
-                _entries[key] = value;
+                auto inserted = _entries.insert(std::pair<QualifiedName, std::shared_ptr<const ObservableEntry>>(key, value));
+                if (!inserted.second)
```
Comment (dvandyk) [inline eos/observable.cc:92]: Coding style: ``! inserted.second``. I wonder if we should rather return the boolean value, rather than ``void`` here, and let the callers of this ``ObservableEntries::insert()`` deal with potential exceptions.

File: eos/observable.cc [hunk 42]
```diff
@@ -292,8 +304,34 @@ namespace eos
     }
 
     void
-    Observables::insert(const QualifiedName & name, const ObservableEntry * entry) const
+    Observables::insert(const QualifiedName & name, const std::string & latex, const Options & forced_options, const std::string & input_expression) const
```
Comment (dvandyk) [inline eos/observable.cc:307]: Hm, the naming is a bit weird. ``input_expression`` is not an expression yet. How about ``input`` and ``expression`` in lieu of ``input_expression`` and ``parsed_expression``? Make the code a bit more readable I think.

File: eos/utils/expression-parser.hh [hunk 43]
```diff
@@ -0,0 +1,59 @@
...
+#ifndef EXPRESSION_PARSER_HH
+#define EXPRESSION_PARSER_HH 1
+
+#include <boost/spirit/include/qi.hpp>
+#include <boost/spirit/include/phoenix.hpp>
+#include <boost/fusion/adapted.hpp>
+
+#include <eos/utils/expression.hh>
+
+namespace qi    = boost::spirit::qi;
```
Comment (dvandyk) [inline eos/utils/expression-parser.hh:27]: Ah, this is dangerous I think, and I overlooked it so far. Can you please move these three namespace declarations inside the ``eos`` namespace?

File: eos/utils/expression-parser_TEST.cc [hunk 44]
```diff
@@ -0,0 +1,200 @@
...
+        }
+};
+
+
+class ExpressionParserTest :
+    public TestCase
+{
+public:
+    ExpressionParserTest() :
+    TestCase("expression_parser_test")
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:60]: Coding style: initalizations/constructor calls are indented one level (4 spaces) with respect to the class name.

File: eos/utils/expression-parser_TEST.cc [hunk 45]
```diff
@@ -0,0 +1,200 @@
...
+    {
+    }
+
+    virtual void run() const
+    {
+        // testing parser failure
+        {
+            ExpressionTest test("1 /* 2");
+
+            TEST_CHECK(!test.completed);
```
Comment (dvandyk) [inline eos/utils/expression-parser_TEST.cc:70]: Coding style: ``! test.completed``

File: eos/utils/expression-visitors.cc [hunk 46]
```diff
@@ -0,0 +1,338 @@
...
+                {
+                    _parameter_user->uses(*observable);
+                }
+
+                return ObservableExpression(observable, e.kinematics_specification);
+            }
+
+            Expression visit(const ObservableExpression & e)
+            {
+                return ObservableExpression(e);
```
Comment (dvandyk) [inline eos/utils/expression-visitors.cc:268]: Ahh, bug! The return value will not necessarily be bound to the same Parameters (``_parameters``). I think we need to clone the existing observable, using ``_parameters`` as the argument to clone. Maybe you were right in the first place that this case should return an error...

File: eos/utils/one-of.hh [hunk 47]
```diff
@@ -189,7 +189,11 @@ namespace eos
             OneOf(const OneOf & other) :
                 _value(other._value)
             {
+            }
 
+            bool empty() const
+            {
+                return !_value;
```
Comment (dvandyk) [inline eos/utils/one-of.hh:196]: Can you split this into a different commit please?

**Review by dvandyk — CHANGES_REQUESTED** [review]

**Comment by mreboud** [comment]
@dvandyk I rebased on master. Unless you have further comments, I suggest we merge this round and open a new issue for the future improvements, including:
 - Manage units in inserted observable entries
 - Ensure that expression observables are cacheable
 - Allow expression observable to be also added at compile time

**Comment by dvandyk** [comment]
> @dvandyk I rebased on master. Unless you have further comments, I suggest we merge this round and open a new issue for the future improvements, including:
> 
> * Manage units in inserted observable entries
> * Ensure that expression observables are cacheable
> * Allow expression observable to be also added at compile time

Sounds good!
